### PR TITLE
update catalog.py

### PR DESCRIPTION
**Pull request by Devhub01**
edit/delete of domains, authors, tags

### CODE REVIEW

File: templates/catalog/papers-in-domain.html [hunk 1]
```diff
@@ -1,5 +1,33 @@
 {% extends "catalog/catalog.html" %}
 {% block title %} Domain {{domainname}} – {% endblock %}
 {% block before_list %}
-  <h4>Papers in domain <em>{{domainname}}:</em></h4>
-{% endblock %}
+
+
+
+<h4>Papers in domain <em>{{domainname}}</em></h4>
```
Comment (kerzol) [inline templates/catalog/papers-in-domain.html]: I propose to use the following code here. It will be more concise.
If you are ok with this modification, could you also change the related templates for tags and authors in the same manner?

```html
<h4>Papers in domain <em>{{domainname}}</em>
{% if user_authenticated() %}
 {% if entry.domainname | can_edit_domain %}
  [<a title="Delete the domain" onclick="return confirm('You really want to delete the domain ?')"
    href="/domain/delete/{{domainname}}">remove</a> /
  <a title="Modify the domain name  " href="/domain/edit/{{domainname}}">edit</a>]
 {% endif %}
{% endif %}
</h4>
```

File: papersite/catalog.py [hunk 2]
```diff
@@ -1,132 +1,365 @@
...
+from papersite.user import (get_user_id, is_super_admin, is_author_of_paper,
+                            is_author_of_comment, user_authenticated, ANONYMOUS)
+
 
 def render_catalog(template_name, **context):
-    domains=query_db ("select * from domains order by domainname")
-    keywords=query_db ("select * from keywords order by keyword")
-    authors=query_db ("select * from authors order by fullname")
-    users=query_db ("select * from users where valid = 1 or userid = 1 order by username")
+    domains = query_db("select * from domains order by domainname")
```
Comment (kerzol) [inline papersite/catalog.py:15]: It is ok for this time. But I strongly advise you to only include in the pull request only the "real" changes. The syntax improvements should go to another pull request.

File: papersite/catalog.py [hunk 3]
```diff
@@ -1,132 +1,365 @@
...
-    else: 
+                         order by title", [q])
+    else:
         papers = []
     return render_catalog('catalog/catalog-search.html',
-                           papers=with_description(papers))
+                          papers=with_description(papers))
+
+
+####################### My Modifications ##############################
```
Comment (kerzol) [inline papersite/catalog.py]: The line is useful for this review. But it can disturb the future reader, because "my" may be related to anyone who participated in the project :)

File: papersite/catalog.py [hunk 4]
```diff
@@ -1,132 +1,365 @@
...
+            error = 'Please enter a valid name'
+
+        else:
+            con = get_db()
+            with con:
+                con.execute('update domains set domainname = ? \
+                                    where domainid = ?',
+                            [request.form['name'], request.form['id']])
+
+                # TODO:
```
Comment (kerzol) [inline papersite/catalog.py]: TODO what ?

File: papersite/catalog.py [hunk 5]
```diff
@@ -1,132 +1,365 @@
...
+
+                flash('You successfully modified the domain')
+                return redirect(url_for('domain',
+                                        domainname=request.form['name']))
+    return render_template('catalog/edit.html',
+                           entry=domain,
+
```
Comment (kerzol) [inline papersite/catalog.py]: Please, remove that empty line

File: papersite/catalog.py [hunk 6]
```diff
@@ -1,132 +1,365 @@
...
+            error = 'Please enter a valid name'
+
+        else:
+            con = get_db()
+            with con:
+                con.execute('update keywords set keyword = ? \
+                                    where keywordid = ?',
+                            [request.form['name'], request.form['id']])
 
+                # TODO:
```
Comment (kerzol) [inline papersite/catalog.py]: same remark as before

File: papersite/catalog.py [hunk 7]
```diff
@@ -1,132 +1,365 @@
...
+            error = 'Please enter a valid name'
+
+        else:
+            con = get_db()
+            with con:
+                con.execute('update authors set fullname = ? \
+                                    where authorid = ?',
+                            [request.form['name'], request.form['id']])
+
+                # TODO:
```
Comment (kerzol) [inline papersite/catalog.py]: same remark as before

File: papersite/db.py [hunk 8]
```diff
@@ -141,19 +154,23 @@ def get_insert_author(fullname):
...
 
 
+#### Updated to check domain existence in DB 
 def get_insert_domain(domainname):
     con = get_db()
     with con:
         con.execute("INSERT OR IGNORE INTO domains(domainname)    \
-                     VALUES(?)", [domainname])
+                    SELECT (?) WHERE not exists                     \
+                    (select (?) from domains where domainname = (?) COLLATE NOCASE)",(domainname,domainname,domainname))
```
Comment (kerzol) [inline papersite/db.py:167]: Bien vu! Superb!

File: templates/catalog/edit.html [hunk 9]
```diff
@@ -0,0 +1,134 @@
+{% extends "layout.html" %}
+{% block bigcontainerclass %}white{% endblock %}
+
+{% block scripts %}
+
+<script>
```
Comment (kerzol) [inline templates/catalog/edit.html:5]: I'm not sure that the autocomplete list is useful here.
Imagine you have two domains named respectively D1 and D2. 
As the user edits the first domain, he can write D2 instead of D1.
And this will generate the error:
u2

It happens because domain (tag, authors) names are UNIQUE in DB.

uni

Moreover, it may be useful to forbid already existing names.
Also, I think that we should not allow the use of a comma "," symbol inside a domain/tag/author name. Because the comma is used to delimiter two different names when someone upload or edit a paper.

File: papersite/catalog.py [hunk 10]
```diff
@@ -1,132 +1,365 @@
...
+# Deleting/Modifying the domains
+#############################
+
+@app.route('/domain/delete/<string:domainname>', methods=['GET'])
+def delete_domain_with_check(domainname):
+    if can_delete_domain(domainname):
+        dom = query_db("select * from domains where domainname = ?",
+                       [domainname], one=True)
+        delete_domain(domainname)
+        delete_papers_domain(dom['domainid'])
```
Comment (kerzol) [inline papersite/catalog.py:203]: It looks a little bit strange to delete a domain (or a tag, or an author's name) when there exist papers that use this domain. Sure, it is powerful, but potentially dangerous operation.
An admin can accidentally remove the domain together with all paper-domain relations !
It seems to me that we should not allow the deletion of a domain (or a tag, or an author's name) if there exist some papers linked to it. Or at least display a warning, before proceeding 

 What do you think ?

**Review by kerzol — CHANGES_REQUESTED** [review]
It works and works well ! 

Below I give my comments related to the user interface style, code style and some comments on the functional part.

Good gob!

File: templates/catalog/edit.html (hunk 9 above)
Comment (Devhub01) [reply templates/catalog/edit.html:5]: I fixed this by deleting the autocomplete function, and testing the existence of the domain using JavaScript. 
This however made it impossible to modify the domain name if it was written for example in capital letter or vice-versa (kdd=KDD). 
Perhaps, we can add a JS function to transform the text value into for example first capital letter of each word(complex networks to  Complex Networks)  when adding a paper !

File: papersite/db.py (hunk 8 above)
Comment (Devhub01) [reply papersite/db.py:167]: Thank you :D

File: papersite/catalog.py (hunk 3 above)
Comment (Devhub01) [reply papersite/catalog.py]: I modified this to "modification made by my GitHub username", so it will help me keep track of what I did in order to write the PFE report, but you can delete this line when merging the branches.

File: templates/catalog/edit.html (hunk 9 above)
Comment (kerzol) [reply templates/catalog/edit.html:5]: A good idea, but I cannot say that it will be 100% ok to silently capitalize all first letters.
Maybe, when we test for the existence we should all domains except the current ?

Comment (Devhub01) [reply templates/catalog/edit.html:5]: Okay, I'll work on it

**Comment by kerzol** [comment]
ok, If you are ready. I propose to merge it.

**Comment by Devhub01** [comment]
Yes please.

**Comment by kerzol** [comment]
" Only those with write access to this repository can merge pull requests. " Could you please merge it or give me a write access to your repo?

**Comment by kerzol** [comment]
To be clear: I propose to merge develop into master inside your repo, than propagate changes into my repo via pull-request.

**Comment by Devhub01** [comment]
Normally I have added you with full control to the repo last time.
I have sent you a new invitation so you can do the pull request, I'm still not quite familiar with git commands

**Comment by Devhub01** [comment]
I think I did it, I merged the two branches inside my repo, and create it a pull request to your repo.
